### ISSUE TITLE
Detect system include paths when using compile_flags.txt

### ISSUE THREAD

**Issue by StMartin81** [clangd/clangd#1089]
Would it be possible to add detection of system include paths when using compile_flags.txt to configure clangd? The system include paths could be detected if:
* compile_flags.txt is used to configure clangd
* the compiler is given in the `CompileFlags` section in .clangd configuration file: `Compiler: <path to compiler executable>`
* the compiler is white listed when executing clangd: `clangd --query-driver=<path to driver>`

**Comment by HighCommander4** [comment]
My understanding is that querying the compiler driver for include paths is supposed to work as well (that is the main purpose of the `CompileFlags: Compiler:` configuration option). Or does using `compile_flags.txt` somehow break that?

**Comment by StMartin81** [comment]
Small example project:
[clangd_compiler_include.zip](https://github.com/clangd/clangd/files/8407445/clangd_compiler_include.zip)

clangd doesn't use the mechanism to resolve the GCC include paths in this case:
```
I[09:11:51.438] clangd version 14.0.0
I[09:11:51.440] Features: windows
I[09:11:51.440] PID: 56144
I[09:11:51.440] Working directory: c:\repositories\clangd_compiler_include
I[09:11:51.440] argv[0]: C:\Program Files\LLVM\bin\clangd.EXE
I[09:11:51.440] argv[1]: --query-driver=**arm-none-eabi-g*
I[09:11:51.447] Starting LSP over stdin/stdout
I[09:11:51.447] <-- initialize(0)
I[09:11:51.449] --> reply:initialize(0) 1 ms
I[09:11:51.452] <-- initialized
I[09:11:51.452] <-- textDocument/didOpen
I[09:11:51.453] <-- clangd/inlayHints(1)
I[09:11:51.456] --> textDocument/publishDiagnostics
I[09:11:51.457] Loaded compilation database from c:\repositories\clangd_compiler_include\compile_flags.txt
I[09:11:51.457] ASTWorker building file c:\repositories\clangd_compiler_include\src\main.c version 1 with command 
[c:\repositories\clangd_compiler_include]
C:/repositories/dev-tools/gcc-arm-win/gcc-arm-none-eabi-10-2020-q4-major/bin/arm-none-eabi-gcc --target=arm-none-eabi -Werror -Wunused -Wuninitialized -Wall -Wextra -Wmissing-declarations -Wpointer-arith -Wshadow -Wlogical-op -Waggregate-return -Wfloat-equal -Wstrict-aliasing -Wcast-align -Wsign-conversion "-resource-dir=C:\\Program Files\\LLVM\\lib\\clang\\14.0.0" -- "c:\\repositories\\clangd_compiler_include\\src\\main.c"
```

**Comment by HighCommander4** [comment]
Thanks, I can reproduce the issue.

It looks like a bug in how the clangd's CDB abstractions are layered. Clangd first builds a `DirectoryBasedGlobalCompilationDatabase`, then wraps that in a `QueryDriverDatabase`, then wraps that in an `OverlayCDB` which applies adjustments to the compile command.

The handling of the `CompileFlags: Compiler:` key is handled at the `OverlayCDB` level, in the form of an adjustment that replaces the command's `argv[0]` with the value of the `Compiler` key.

However, the querying has already happened at the `QueryDriverDatabase` level. At that time, the adjustment to respect the `Compiler` key has not yet been made, so the `argv[0]` was a generic `clang-tool`. It is this `clang-tool` that clangd tried to query (and bailed because it did not match the `--query-driver` regex).

**Comment by HighCommander4** [comment]
The above diagnosis suggests that the issue is not specific to `compile_flags.txt`, it affects `compile_commands.json` as well (though I guess you're less likely to need to use `CompileFlags: Compiler:` if you're using `compile_commands.json`).

**Comment by danngreen** [comment]
I have the same issue, using `compile_commands.json`.

I'm not sure if this is helpful information, but I would also note that in clangd 13.0.0, the bug is not present.

In both versions 13 and 14, clangd successfully queries the compiler and prints the correct system dirs in the logs.
In clangd 13, the `-isystem <DIR>` flags are appended to the compile command for any file which lacks an entry in the CDB (and which clangd sucessfully "infers" the compile command from another file). However, in clangd 14.0.6, the `-isystem` flags are not present in the compile command found in the log. Thus, we get diagnostics like:

```
#include <cstdint>            <----'cstdint` file found
```

The workaround I've found is to manually paste in `-isystem` flags into a .clangd file.
Post-processing the compile-commands.json file to add an entry for each unknown file would probably work, too, except not in the case where I create a new file and immediately see diagnostic errors when including standard libraries.

In any case, if I can help with a minimal example or anything, I'm happy to.

**Comment by HighCommander4** [comment]
@danngreen Based on your description, you're running into a different issue, #1100, which is fixed on the 15.x branch and trunk.

**Comment by danngreen** [comment]
Ah, yes, that's it exactly. Weekly build fixes it, thanks!

**Comment by HighCommander4** [comment]
Proposed patches:
https://reviews.llvm.org/D133756
https://reviews.llvm.org/D133757

They also fix #1173 and #1263

**Comment by HighCommander4** [comment]
For folks following this issue, just wanted to give a heads up that this has been fixed in clangd 16, though there is a remaining issue discussed in #1568.

The remaining issue affects scenarios where:
 * You're trying to open a file with extension `.h` (such that clangd can't infer the language from the extension alone)
 * Your compiler driver is gcc
 * You do not have gcc's objective-c support installed (see details in #1568 for why this is relevant)

In such scenarios, a workaround for this is to add a `.clangd` file (or add a section to your existing `.clangd` file) containing:

```yaml
If:
  PathMatch: *\.h

CompileFlags:
  Add: [-xc++-header]
```

(or `-xc-header` if you're working on a C project and your `.h` files should be parsed as C).
